### ISSUE TITLE
There are better ways to add the mono compatibility to support Razor views

### ISSUE THREAD

**Issue by davidfowl** [jbatonnet/Microsoft.AspNetCore.Mono#2]
A few things before I get to the main issue. I read your post about ASP.NET Core on the Pi (very cool! https://www.thedju.net/articles/asp-net-core-mono-linux). 

Dependency Model is OSS:
https://github.com/dotnet/core-setup/tree/master/src/Microsoft.Extensions.DependencyModel

You probably shouldn't use private reflection to edit the reference assemblies. I know it works but it's fragile. If you're trying to tell razor for where the compilation assemblies are, there are better extensibility points for that:

https://github.com/aspnet/Mvc/blob/rel/1.1.0/src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngineOptions.cs#L98

You can add to the list of assemblies here so that runtime compilation works. This is a public supported API.

PS .NET Core now works on ARM

**Comment by jbatonnet** [comment]
Hi @davidfowl,

Thank you for your feedback. This is obviously an ugly way to fix ASP.NET Core compatibility on Mono.
I actually had a similar conversation in a pull request on this repo here : https://github.com/jbatonnet/Microsoft.AspNetCore.Mono/pull/1

I prepared a PR for https://github.com/dotnet/core-setup/tree/master/src/Microsoft.Extensions.DependencyModel, but this kind of tricks does not seem to be needed anymore to make ASP.NET Core run on Mono.
I tested it on a Raspberry Pi 2 with Mono 4+, and everything seems to work as expected.
You will still need a replacement for Kestrel as libuv somehow still fail to run.
I now use https://github.com/jbatonnet/Microsoft.AspNetCore.Server.HttpListener instead of EmbedIO to reduce overhead.

This week I saw .NET Core works fine on ARM. I personally think that .NET Core is not mature enough to be used in production. It lacks a lot of feature from .NET Framework, the project.json structure is far less powerful than csproj.
I'm waiting for .NET Standard 2 / VS 2017 to be released to finally use the same code :
- On .NET Framework / Windows
- On Xamarin / Android (already the case)
- On .NET Core / Linux 😎

**Comment by davidfowl** [comment]
I don't have a problem with adding servers to support mono as that's an actual extensibility point. The dependency model hacking is what needs to go.

**Comment by flaub** [comment]
@davidfowl: Assuming you're referring to `AdditionalCompilationReferences` as the officially supported extensibility point, let me show you the kind of situation I run into using this technique to make things work on Mono:
```
mvcBuilder.AddRazorOptions(options =>
{
    var assemblies = new[]
    {
        typeof(BootstrapMvc.Alert).Assembly,
        typeof(BootstrapMvc.Core.IItemWriter).Assembly,
        typeof(BootstrapMvc.Mvc6.BootstrapHelper).Assembly,
        typeof(Microsoft.AspNetCore.Html.IHtmlContent).Assembly,
        typeof(Microsoft.AspNetCore.Http.Features.IFormFeature).Assembly,
        typeof(Microsoft.AspNetCore.Mvc.IUrlHelper).Assembly,
        typeof(Microsoft.AspNetCore.Mvc.UrlHelperExtensions).Assembly,
        typeof(Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper).Assembly,
        typeof(Microsoft.AspNetCore.Razor.TagHelpers.HtmlAttributeValueStyle).Assembly,
        typeof(Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperRunner).Assembly,
        typeof(System.Object).Assembly,
        typeof(System.Collections.Generic.ISet<int>).Assembly,
        typeof(System.Linq.Enumerable).Assembly,
        typeof(System.Text.Encodings.Web.HtmlEncoder).Assembly,
        AppDomain.CurrentDomain.GetAssemblies().Single(x => x.GetName().Name == "System.Runtime"),
    };

    foreach (var assembly in assemblies)
    {
        options.AdditionalCompilationReferences.Add(
            MetadataReference.CreateFromFile(assembly.Location)
        );
    }
});
```

From a user's point of view, I'd much rather say `UseMonoCompatibility` than deal with the above mess. I frequently have to update this code when a new dependency comes along. I'd be OK if Razor pre-compilation would work, but so far the tooling and mono aren't yet in sync.

**Comment by davidfowl** [comment]
@flaub That's the sort thing you would do in `UseMonoCompatibility`, not hack a static private variable. More specifically, you would add an `IConfigureOptions<RazorOptions>` to the DI container that had the logic to resolve mono reference assemblies for compilation.

**Comment by flaub** [comment]
The thing is, for Mono, I don't want to specify each assembly reference. Instead I want to specify a search path. Ideally, the system would know to look for assemblies in the application path and also the equivalent of the GAC. I agree that hacking a static private variable is no good. But it's also no good that the only extensible thing is at the granularity of assemblies.

**Comment by davidfowl** [comment]
> The thing is, for Mono, I don't want to specify each assembly reference. Instead I want to specify a search path.

Runtime compilation doesn't hardcode any knowledge of the layout of reference assemblies. That's why the toolchain copies the reference assemblies to the output folder for all non mono cases (even .NET Framework doesn't hard code any GAC knowledge). See PreserveCompilationContext.

> Ideally, the system would know to look for assemblies in the application path and also the equivalent of the GAC. I agree that hacking a static private variable is no good. 

ASP.NET Core doesn't and won't hard code knowledge of the GAC. That's why the extensibility does this.

> But it's also no good that the only extensible thing is at the granularity of assemblies.

It's literally the most flexible thing you can expose, any search capability can be built on top of it. The mono reference assembly location varies per OS and based on where you installed mono. Adding a mono aware razor package would be fine here.

**Comment by flaub** [comment]
Fair enough, I'll start prototyping a method for adding a search capability thru this interface.

Do you happen to know of any code within .NET (not necessarily .NET Core) that is shared/open that has a good chance of already providing the proper search algorithm? I suppose _proper_ is somewhat context-specific. I'd really just like to use whatever algorithm the pre-compilation system is using. Ideally we'd allow users to get the same same semantics as the pre-compilation system except it happens at runtime.  Perhaps that's what you meant by the reference to the PreserveCompilationContext.
